### PR TITLE
fix: add report user button back to profile pages

### LINKED ISSUE #36837: No way to report users
There used to be a button to report users:
https://github.com/freeCodeCamp/freeCodeCamp/blob/46a217d0a5b4f6b499ce24a18b46fd78fff45d78/api-server/common/app/routes/Profile/Profile.jsx#L190
added to close https://github.com/freeCodeCamp/freeCodeCamp/issues/16598, but this has since been removed.  However, there's still code associated with user reports:
https://github.com/freeCodeCamp/freeCodeCamp/blob/e54a7fb35097766a187cab1a26e752e4e22e946e/client/src/client-only-routes/ShowUser.js#L136-L139

Do we still want this?  If not, all the associated code can go.

### PR DESCRIPTION
**Pull request by moT01**
As far as I can tell, this seems to work. It will put a `Report This User` button in place of the settings button anytime it's not the users own profile.

From there it will bring up the report form for an authenticated user, and redirect a non authenticated user to `/signin` after 5 seconds

~~I was unable to actually send a report email to myself, but I'm guessing that will work in production?~~ Looks like mailhog did catch those - didn't know to check there, so this should be good.

~~Also, the automatic redirect to `/signin` doesn't work locally - I'm pretty sure that will be fine on production.~~ - this should be fixed now

~~If someone can confirm those last two, I will open this for review.~~

- [x] I have read [freeCodeCamp's contribution guidelines](https://github.com/freeCodeCamp/freeCodeCamp/blob/master/CONTRIBUTING.md).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `master` branch of freeCodeCamp.
- [x] None of my changes are plagiarized from another source without proper attribution.
- [x] All the files I changed are in the same world language (for example: only English changes, or only Chinese changes, etc.)
- [x] My changes do not use shortened URLs or affiliate links.

Closes #36837

### CODE REVIEW

**Comment by raisedadead** [comment]
> From there it will bring up the report form for an authenticated user, and redirect a non authenticated user to `/signin` after 5 seconds

Is this an intended wait time somewhere? Or you are proposing this?

> Also, the automatic redirect to `/signin` doesn't work locally - I'm pretty sure that will be fine on production.

This should work on local as well. There is no reason for it to break. We have mocked the OAuth 2.0 strategy. You will not see a prompt but the api logs on your terminal should show up hitting the paths.

Anyways, to be in the clear, you should be able to setup Auth0 and get your own keys. This is particularly useful if you will be working with authenticated routes very often.

I can hook you up with the keys if you need to be very close to production. But the type of auth (passwordless/social, etc.) really does not matter. The defaults you get from Auth0 (password based) should be fine as it is. The API will see it as a OAuth 2.0 provider.

Setting this on the Auth0 Application setting should be fine (everything else can be ignored):

```
Allowed Callback URLs: http://localhost:3000/auth/auth0/callback
```

Set these:

```
AUTH0_CLIENT_ID=stuff
AUTH0_CLIENT_SECRET=stuff
AUTH0_DOMAIN=stuff
```

Disable local mocking

```
LOCAL_MOCK_AUTH=false
```

**Comment by moT01** [comment]
> From there it will bring up the report form for an authenticated user, and redirect a non authenticated user to /signin after 5 seconds

This was the functionality that was already in place on the [showUser.js file](https://github.com/freeCodeCamp/freeCodeCamp/blob/master/client/src/client-only-routes/ShowUser.js) - I assume that was the page that users were taken to for reporting in the past. I just added the button to get there.

**Comment by raisedadead** [comment]
> This was the functionality that was already in place [...]

Thanks for pointing me to this :) Let's leave that as is.

**Comment by moT01** [comment]
I think I figured out the local sign in issues when comparing that file to other redirects to /signin. I added a commit which fixes it for local redirects.

**Comment by moT01** [comment]
The last commit there is for when you submit a report, it redirects to learn instead of the homepage because the homepage automatically redirects to learn, so there was like a double redirect that was annoying.

**Comment by raisedadead** [comment]
Nice work @moT01. 
Can we add some tests to go with this, because we are fixing a regression?

File: client/src/client-only-routes/ShowUser.js [hunk 1]
```diff
@@ -14,8 +13,11 @@ import {
 } from '@freecodecamp/react-bootstrap';
 import Helmet from 'react-helmet';
 
+import { apiLocation } from '../../config/env.json';
+
 import {
   isSignedInSelector,
+  hardGoTo,
```
Comment (ojeytonwilliams) [inline client/src/client-only-routes/ShowUser.js]:
```suggestion
  hardGoTo as navigate,
```
If we alias like this, we can clean up a little.  See below.

File: client/src/client-only-routes/ShowUser.js [hunk 2]
```diff
@@ -24,7 +26,9 @@ import { Spacer, Loader, FullWidthRow } from '../components/helpers';
 
 const propTypes = {
   email: PropTypes.string,
+  hardGoTo: PropTypes.func.isRequired,
```
Comment (ojeytonwilliams) [inline client/src/client-only-routes/ShowUser.js]:
```suggestion
```

File: client/src/client-only-routes/ShowUser.js [hunk 3]
```diff
@@ -46,7 +50,14 @@ const mapStateToProps = createSelector(
 );
 
 const mapDispatchToProps = dispatch =>
-  bindActionCreators({ reportUser }, dispatch);
+  bindActionCreators(
+    {
+      hardGoTo,
```
Comment (ojeytonwilliams) [inline client/src/client-only-routes/ShowUser.js]:
```suggestion
```

File: client/src/client-only-routes/ShowUser.js [hunk 4]
```diff
@@ -46,7 +50,14 @@ const mapStateToProps = createSelector(
...
+      navigate: location => dispatch(hardGoTo(location)),
```
Comment (ojeytonwilliams) [inline client/src/client-only-routes/ShowUser.js]:
```suggestion
      navigate,
```

**Review by ojeytonwilliams — CHANGES_REQUESTED** [review]
Looks great, @moT01.

The only thing was that `navigate` was essentially an alias for `hardGoTo`, so I made some suggestions to make that explicit.

Regardless, everything is working nicely.

**Comment by ojeytonwilliams** [comment]
@raisedadead unless I'm missing something, `hardGoTo` achieves the same thing as reach-router's `navigate`.  Is there a reason to keep it or can we just cut it out?  In general, that is, rather than just this PR.

**Comment by raisedadead** [comment]
> Is there a reason to keep it or can we just cut it out? 

I think it was a legacy thing that we had before we had `navigate`. It can be removed and cleaned up.

File: client/src/client-only-routes/ShowUser.js [hunk 5]
```diff
@@ -1,7 +1,6 @@
 import React, { Component, Fragment } from 'react';
 import PropTypes from 'prop-types';
 import { bindActionCreators } from 'redux';
```
Comment (ojeytonwilliams) [inline client/src/client-only-routes/ShowUser.js]:
```suggestion
import { bindActionCreators } from 'redux';
import { navigate } from '@reach/router';
```
https://github.com/freeCodeCamp/freeCodeCamp/pull/36926#issuecomment-537857201

**Review by ojeytonwilliams — APPROVED** [review]
LGTM :+1:

**Review by raisedadead — APPROVED** [review]

**Pull request merged**